plugins/aggregators/derivative/derivative.go

```diff
@@ -0,0 +1,234 @@
+	## such roll-overs can be configured with a default of 10.
+	# max_roll_over = 10
+	##
+`
+
+func (d *Derivative) SampleConfig() string {
+	return sampleConfig
+}
+
+func (d *Derivative) Description() string {
+	return "Calculates a derivative for every field."
+}
+
+func (d *Derivative) Add(in telegraf.Metric) {
+	id := in.HashID()
+	if current, ok := d.cache[id]; !ok {
+		// hit an uncached metric, create caches for first time:
+		d.cache[id] = newAggregate(in)
+	} else {
+		if current.first.time.After(in.Time()) {
+			current.first = newEvent(in)
+			current.rollOver = 0
+		}
+		if current.first.time.Equal(in.Time()) {
+			upsertConvertedFields(in.Fields(), current.first.fields)
+			current.rollOver = 0
+		}
+		if current.last.time.Before(in.Time()) {
+			current.last = newEvent(in)
+			current.rollOver = 0
+		}
+		if current.last.time.Equal(in.Time()) {
+			upsertConvertedFields(in.Fields(), current.last.fields)
+			current.rollOver = 0
+		}
+
+		d.cache[id] = current
+	}
+}
+
+func newAggregate(in telegraf.Metric) aggregate {
+	return aggregate{
+		name:     in.Name(),
+		tags:     in.Tags(),
+		first:    newEvent(in),
+		last:     newEvent(in),
+		rollOver: 0,
+	}
+}
+
+func newEvent(in telegraf.Metric) *event {
+	return &event{
+		fields: extractConvertedFields(in),
+		time:   in.Time(),
+	}
+}
+
+func extractConvertedFields(in telegraf.Metric) map[string]float64 {
+	fields := make(map[string]float64, len(in.Fields()))
+	upsertConvertedFields(in.Fields(), fields)
+	return fields
+}
+
+func upsertConvertedFields(source map[string]interface{}, target map[string]float64) {
+	for k, v := range source {
+		if value, ok := convert(v); ok {
+			target[k] = value
+		}
+	}
+}
+
+func convert(in interface{}) (float64, bool) {
+	switch v := in.(type) {
+	case float64:
+		return v, true
+	case int64:
+		return float64(v), true
+	case uint64:
+		return float64(v), true
+	default:
+		return 0, false
+	}
+}
+
+func (d *Derivative) Push(acc telegraf.Accumulator) {
+	for _, aggregate := range d.cache {
+		if aggregate.first == aggregate.last {
+			d.Log.Debugf("Same first and last event for %q, skipping.", aggregate.name)
+			continue
+		}
+		var denominator float64
+		if len(d.Variable) == 0 {
+			// If no derivation variable is known default to timestamp
+			denominator = aggregate.last.time.Sub(aggregate.first.time).Seconds()
+		} else {
+			var first float64
+			var last float64
+			var found bool
+			if first, found = aggregate.first.fields[d.Variable]; !found {
+				d.Log.Infof("Did not find %q in first event for %q.", d.Variable, aggregate.name)
+				continue
+			}
+			if last, found = aggregate.last.fields[d.Variable]; !found {
+				d.Log.Infof("Did not find %q in last event for %q.", d.Variable, aggregate.name)
+				continue
+			}
+			denominator = last - first
+		}
+		if denominator == 0 {
+			d.Log.Infof("Got difference 0 in denominator for %q, skipping.", aggregate.name)
```

Comment:
Not sure if `Info` is a sufficient logging level as we drop metrics. It is a balance between likelihood and severeness though. Please rethink if this shouldn't be `Warn`. But as I do not have any data, you should decide here.

Reply:
This needs some more work. This code branch is also taken, when there was only one metric during the period. This needs to be changed, before changing the log level to warn. I will work on that.

Reply:
Actually, when there is only one metric in the period, calculation of the period is also skipped. Hence, a higher log level would be needed according to your suggestion. I will change all log levels to WARN and fix the initialisation of a new aggregate.